### PR TITLE
Isaicp 4300: API to identify schema fields

### PR DESCRIPTION
**Pull request by idimopoulos**

### CODE REVIEW

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    $search = ['@graph', '@property_predicates', '@rdf_type', '@field_iri'];
+    $replace = [
+      SparqlArg::uri($graph),
+      SparqlArg::serializeUris($property_predicates, ' '),
+      SparqlArg::uri($rdf_type),
+      SparqlArg::uri($field_iri),
+    ];
+
+    $query = <<<QUERY
+ASK {
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: I'm trying to understand the query. This reads:

Check if there is a schema entry in `http://adms-definition` where ALL next conditions are met:
1. The schema entry has subject (`?entity_id`) equals to the Drupal field column mapping.
1. The entry has the same type/range (`rdf:type`/`a`) as the Drupal field.
1. The entry belongs to a domain that is mapped to the field's bundle.

Is this correct? Please explain if not (I'm just curious)

File: web/modules/custom/rdf_ff/rdf_ff.module [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+/**
+ * @file
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/rdf_ff.module]: Php codesniffer finds this as an error though.

```
FILE: ...m/rdf_schema_field_validation/rdf_schema_field_validation.module
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 1 | ERROR | [x] Missing file doc comment
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+   *
+   * @param string $entity_type_id
+   *   The entity type id.
+   * @param string $bundle
+   *   The entity type id.
+   *
+   * @return \Drupal\rdf_entity\RdfEntityMappingInterface|null
+   *   The rdf entity mapping or null, if none is found.
+   */
+  protected function getEntityMapping($entity_type_id, $bundle): ?RdfEntityMappingInterface {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: I was taught by Pieter that we should not use static methods in OOP as it is hard to mock the classes (if I am not mistaking). this is there because I am using the passed entity type manager instead of the global one.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+   *   The bundle of the class defined in the ontology.
+   * @param string $field_name
+   *   The field name.
+   * @param string $column
+   *   The column name.
+   *
+   * @return bool
+   *   Whether the field name is defined in the schema.
+   */
+  public function isDefinedInSchema($entity_type_id, $bundle, $field_name, $column = '');
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php]: Well we the `\Drupal\rdf_entity\RdfFieldHandler::getFieldPredicates` method already does that. If no column is passed, it fetches the main property.
We don't need actually a per field check because the field is not part of the schema. Only the columns are.

Maybe we need something like that when the blank nodes are implemented but I think it will take some time for that to happen :/

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php]: Also, isDefinedInSchema is better now that I have renamed the module :D

File: web/modules/custom/rdf_ff/tests/src/Kernel/RdfFieldSchemaValidationTest.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    $query = "INSERT DATA INTO $graph_uri {\n";
+    $query .= $graph->serialise('ntriples') . "\n";
+    $query .= '}';
+    $this->spaqlEndpoint->update($query);
+  }
+
+  /**
+   * Asserts that fields belong or don't belong to a defined schema.
+   */
+  public function testFieldSchema() {
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/tests/src/Kernel/RdfFieldSchemaValidationTest.php]: The idea here is that we are using adms but the module simply tests fields defined in schema. If the schema is correct, then proper checks occur. 
The idea was to create a companion module to the rdf_entity which is joinup independent. I was told to "do it right". 
If it was a joinup specific, we could even set a static list. This is the reason I also got into the trouble with the 3rd party settings and the schema.

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/tests/src/Kernel/RdfFieldSchemaValidationTest.php]: Maybe we could also have a test on joinup like this as well.

File: web/modules/custom/rdf_ff/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Rdf federated fields
+
+This simple module provides a service that validates that a field
+belongs to an ontology.
+
+This can be used for entities that have mapped rdf properties 
+using the `rdf_entity_mapping` entity type.
+
+It provides some third party settings to that entity type that stores
```
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/README.md]: I answered above on that. My thoughts were that the module is a very small joinup-independent module that can accompany rdf_entity. In that sense, each rdf_entity defined, might have a different type, e.g. a collection defined in ADMS, something else defined entirely in DCAT, another defined in another specification, etc. In the same sence, the predicates that connect the parent class and the properties, might differ.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 1 above)
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: It was wrong. I updated it an placed a comment as well.

File: web/modules/custom/rdf_ff/rdf_ff.module (hunk 2 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/rdf_ff.module]: The entire empty file should be deleted not only the the docblock.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: Hm, I heard this too from him but I have to disagree. The static call, calls also the entity type manager. In both cases you'll need to mock the EntityTypeManager. The static call will work from unit test because that calls the mocked manager. Also, I found no documentation related to this, no recommendation, no "best practices". I think this deserves, eventually to be well documented somewhere and for this reason I've opened an upstream issue: https://www.drupal.org/project/drupal/issues/2945539. But till I don't have an "official" point of view from the community I'm keeping my point of view, that the static call shorthand is correct used anywhere.

File: web/modules/custom/rdf_ff/rdf_ff.info.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+name: 'Rdf federated fields'
```
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/rdf_ff.info.yml]: Now I see that is `rdf_schema_field_validation`. Why "schema field validation", I don't get it? But I'm not a native English speaker so I may be in mistake. For me sounds more as "field schema validation". 

Also the human-readable name should follow the new name, no?

File: web/modules/custom/asset_release/asset_release.info.yml [hunk 8]
```diff
@@ -19,6 +19,7 @@ dependencies:
   - options
   - owner
   - rdf_entity
+  - rdf_ff
```
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/asset_release/asset_release.info.yml]: I see you removed the `rdf_ff` but `asset_distribution` was not added (?)

File: web/modules/custom/rdf_schema_field_validation/README.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+The module needs that the ontology definition is stored in a graph
+and this graph is stored in the third party settings of the
+corresponding `rdf_entity_mapping` entity of the entity type in question.
+The rdf property that defines relationships for example, is 
+`<http://www.w3.org/2000/01/rdf-schema#domain>`.
+
+## Usage
+```
+$entity_type_id = 'entity_test';
+$bundle = 'bunlde';
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/README.md]: s/'bunlde'/'bundle' :)

File: web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php]: > We don't need actually a per field check because the field is not part of the schema.

This issue is about providing an API to "Identify if an individual field is an ADMS field" (issue title). Moreover, as I understood, we want, in a future ticket, to block the edit of such fields. But from a Drupal perspective the unit is the field. The access is granted on a field level, not a column level. Meaning that, when we'll have to implement the per-field access, we'll have to do that logic there (iterating over columns or picking the main column). But as I told, this is the scope of this ticket: "Give me entity type, bundle and field name and I'll tell you if it's an ADMS field". Still think that we need a per-field check in order to support the future ticket.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  public function __construct(Connection $sparql_endpoint, EntityTypeManagerInterface $entity_type_manager, RdfFieldHandler $field_hanlder) {
+    $this->sparqlEndpoint = $sparql_endpoint;
+    $this->entityTypeManager = $entity_type_manager;
+    $this->fieldHanlder = $field_hanlder;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Exception
```
[the remark this answers is not in the document]
Comment (pfrenssen) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: Yep, we cannot deviate from the documentation defined in the public interface, since this is the public "contract" that we are fulfilling. We shouldn't throw exceptions here if this is not defined in the interface.

If you do need to throw exceptions then an acceptable workaround is to throw a `\RuntimeException`. These are problems that are not bugs in the calling code but are happening because of problems in runtime (e.g. if we lose the connection to a remote database this is a runtime error). Runtime exceptions don't need to be documented.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: First of all, from an OO perspective we should always use injected dependencies and not rely on hardcoded implementations. In OOP a class is intented to be a reusable object, and any dependencies it has should be declared through interfaces. This makes it architecture agnostic. This OOP pattern is called [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
In fact the dependency injection framework we use in Drupal is even from another architecture, we are using the [Symfony service container](https://symfony.com/doc/3.4/service_container.html) component. Our services should be exchangeable with services written for Symfony with little to no adaptations. If we call `RdfEntityMapping::load()` in our code the class suddenly becomes hard coupled to the Drupal 8 implementation of entities, and is not fully reusable. What is particularly bad about it is that it hard couples to an _implementation_ (`\Drupal\rdf_entity\Entity\RdfEntityMapping` in full). If it was a static method defined on an interface it wouldn't even be _that_ bad, but by specifying the implementation itself we can no longer override it, we are married to that specific implementation.

Secondly, the one and only reason that we have `EntityInterface::load()` and `EntityInterface::loadMultiple()` is because this made the task of porting core and contrib from Drupal 7 to Drupal 8 vastly easier. They are intended to be used as direct replacements for `entity_load()` and `entity_load_multiple()` (and their entity specific cousins like `node_load()` and friends) _in procedural code_.  It was never intended to be used in OO classes.

At the time this was introduced (ref. https://www.drupal.org/node/2190313) we were in the middle of porting legacy Drupal 7 code and were faced with replacing tens of thousands of instances of `entity_load()` with `\Drupal::entityManager()->getStorage($entity_type_id)->load($entity_id)`. We were concerned about the bad DX of this pattern and people were actively complaining that D8 code was too hard to write.

Thirdly, this makes it unnecessarily difficult to unit test. In a proper unit test only the class itself (the "System under test") is loaded, and all other dependencies are represented by controlled mocks. If we call `RdfEntityMapping::loadByName()` then suddenly this class gets loaded and the method executed, which is something we don't want in a controlled unit test. To make it worse, in turn this code will load the entire `\Drupal` class statically. This continues until it finally throws a `ContainerNotInitializedException` because in a unit test the container doesn't exist. I've seen some unit tests that work around this creatively, but if we simply follow the standard OOP pattern then we can simply mock the entity storage properly.

File: web/modules/custom/rdf_ff/rdf_ff.services.yml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+services:
+  rdf_ff.schema_field_validator:
```
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/rdf_ff.services.yml]: This works now.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: @pfrenssen could we have this opinion copied in the upstream ticket? https://www.drupal.org/project/drupal/issues/2945539. I want to answer there because, finally, we want this written somewhere as being (or not) a pattern that we should follow.

File: web/modules/custom/solution/solution.info.yml [hunk 12]
```diff
@@ -16,6 +16,7 @@ dependencies:
   - options
   - owner
   - rdf_entity
+  - rdf_ff
```
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/solution/solution.info.yml]: OK. But don't forget the `asset_release` dependency on `asset_distribution`.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (pfrenssen) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: I also posted a reply in the upstream issue: https://www.drupal.org/project/drupal/issues/2945539#comment-12489703

[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: OK, thanks, I'll reply there. DX is dead :)

File: web/modules/custom/rdf_ff/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/README.md]: Yes, it's "very small" but creates some unneeded complexity. Let's keep it in this way if this was requested from Sander. He knows better than me.

File: web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  $gs = new GraphStore($connect_string);
+  $graph = new Graph();
+  try {
+    $graph->parseFile(DRUPAL_ROOT . '/../resources/fixtures/adms-definition.rdf');
+  }
+  catch (\Exception $e) {
+    throw $e;
+  }
+  $out = $gs->replace($graph, $graph_uri);
+  if (!$out->isSuccessful()) {
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: This is useless. `GraphStore::replace()` is already throwing an error if `isSuccessful()` returns `FALSE`. See https://github.com/njh/easyrdf/blob/master/lib/GraphStore.php#L153

Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: This is useless. GraphStore::replace() is already throwing an error if isSuccessful() returns FALSE. See https://github.com/njh/easyrdf/blob/master/lib/GraphStore.php#L153

File: web/modules/custom/rdf_schema_field_validation/src/SchemaFieldValidatorInterface.php [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+   *   The bundle of the class defined in the ontology.
+   * @param string $field_name
+   *   The field name.
+   * @param string $column
+   *   The column name.
+   *
+   * @return bool
+   *   Whether the field name is defined in the schema.
+   */
+  public function isDefinedInSchema(string $entity_type_id, string $bundle, string $field_name, string $column = '');
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/src/SchemaFieldValidatorInterface.php]: Lacks the return type :)

File: web/modules/custom/rdf_ff/rdf_ff.info.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/rdf_ff.info.yml]: I was thinking that the 'rdf schema' is the category and the 'field validation' is the reason of the module. Like rdf_entity_mapping is for rdf_entity, which is the 'area' and the mapping which is the actual reason.

[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/rdf_ff.info.yml]: But anyway I don't mind, I can change it if you want.

File: web/modules/custom/asset_release/asset_release.info.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/asset_release/asset_release.info.yml]: Because it is not called 'rdf_ff' anymore :)

File: web/modules/custom/rdf_ff/rdf_ff.module (hunk 2 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/rdf_ff.module]: ahh. Removed.

File: web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php]: The problem with what you are saying is the definition of the terms.
We are bridging the two database systems but the idea does not change.

In drupal we have fields. These fields have properties. For example, the text field has the properties `value`, `format`, the integer field has the `value`, `prefix`, `suffix` etc.
These are stored in the `field_data_field_name` table. 

However, in the triplestore, we do not have such thing. We have blank nodes but they are not supported at the moment, and even if they were, we do not map properties to fields(this is the reason why we do not support deltas at the moment).
So a graph triple will not contain a field value, which is a group of properties, rather than a single column of the field.

In addition to that, we have made some compromises to manage and store the complexity of drupal fields to the triplestore. An example of that is that for text fields, the language is stored as a suffix. E.g., the English version of a French word "Liberation" will be stored as `"Liberty"@en` because in triplestore, the language for strings, is stored as a suffix.
For other cases, it might be an independent property.

For these reasons, we cannot track a field as an entity but we have to track each property separately.
The solution to that issue comes by the fact that we take care and store each main adms field into a different Drupal field. So we do not have 2 ADMS fields as properties of the same Drupal field.

Later on, when we need to check if a field of an entity type and a bundle are part of the ADMS schema, we need to call the method `::isDefinedInSchema` without a column passed. The method will then use the main property as a value which can include a predicate included in the ADMS.

In conclusion, we do not store a relation between a field and a schema value. If you see any of the field.storage related to the rdf entities or even the rdf_entity mappings that you created earlier, none of the predicates are mapped to a field or an entity in general. They are mapped to properties.

The functionality is meant to check if a Drupal property (whether a base field or an extra field property) matches an ADMS defined property. 
If a property is not passed, the default property is used. Currently the default properties are these that we use to match the ADMS fields.

**Comment by idimopoulos** [comment]
The failures in the tests were due to new tests introduced by the epic that should have the rdf_schema_field_validation module enabled now as they are checking joinup entities.

File: web/modules/custom/rdf_ff/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (idimopoulos) [reply web/modules/custom/rdf_ff/README.md]: This part was not requested by Sander. This third party thing is there to provide a flexibility for the entities. It is built now anyway and we don't need to hardcode urls. Also, as far as I remember, the team was always against hardcodings.

File: web/modules/custom/rdf_ff/rdf_ff.info.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/rdf_ff.info.yml]: OK. Let's keep it in this way. Just that somehow confused me.

[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/rdf_ff.info.yml]: Wait, the human-readable name still not fixed :)

File: web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+use Drupal\rdf_entity\Entity\RdfEntityMapping;
+use EasyRdf\Graph;
+use EasyRdf\GraphStore;
+
+/**
+ * Implements hook_update_N().
+ */
+function rdf_schema_field_validation_update_8102() {
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: This will never run. If you want to decouple that from the module (because the module should not be aware about that specific implementation) you should move it in core module.

File: web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+  $connection_options = $sparql_endpoint->getConnectionOptions();
+  $connect_string = 'http://' . $connection_options['host'] . ':' . $connection_options['port'] . '/sparql-graph-crud';
+  // Use a local SPARQL 1.1 Graph Store.
+  $gs = new GraphStore($connect_string);
+  $graph = new Graph();
+  try {
+    $graph->parseFile(DRUPAL_ROOT . '/../resources/fixtures/adms-definition.rdf');
+  }
+  catch (\Exception $e) {
+    throw $e;
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: This is weird :) In fact all this `try {...} catch() {...}` block can be rewritten as 
```php
$graph->parseFile(DRUPAL_ROOT . '/../resources/fixtures/adms-definition.rdf');
```
I know, the IDE. Ignore it

File: web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+  }
+  $gs->replace($graph, $graph_uri);
```
Comment (claudiu-cristea) [inline web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: Tip: You can put the URI in graph directly, few lines above:
```
$graph = new Graph($graph_uri);
```
then
```
$gs->replace($graph);
```

File: web/modules/custom/rdf_ff/src/SchemaFieldValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidator.php]: I'm keeping here my point. I specially created `::loadByName()` as a shorthand, to avoid crating such useless method (in fact is mocking EntityInterface::load()). But let's go with the actual code, I'm OK

File: web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/src/SchemaFieldValidatorInterface.php]: OK. Let's go

File: web/modules/custom/rdf_ff/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (claudiu-cristea) [reply web/modules/custom/rdf_ff/README.md]: OK, Let's keep it.

File: web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install (hunk 16 above)
Comment (idimopoulos) [reply web/modules/custom/rdf_schema_field_validation/rdf_schema_field_validation.install]: No the IDE was not complaining on that. I don't remember why I had the try-catch. I am going to remove it because anyway, I am not cancelling any operation so it does not make sense.
